tools/ir-generator/irclass.cpp

```diff
@@ -193,11 +196,38 @@ void IrDefinitions::generate(std::ostream &t, std::ostream &out, std::ostream &i
     t << std::endl;
 
     t << "namespace IR {" << std::endl;
-    for (auto cls : *getClasses()) {
+
+    // Emit forward declarations
+    for (auto *cls : *getClasses()) {
         enter_namespace(t, cls->containedIn);
         cls->declare(t);
         exit_namespace(t, cls->containedIn);
     }
+
+    t << std::endl;
+
+    // Emit node kinds
+    // TODO: Probably it would make sense to topo-sort the IDs to optimize the
+    // comparison trees generated by a compiler
+    t << "enum class NodeKind : uint64_t {\n"
+      << "  Auto = 0,\n"
+      << "  INode = 1,\n"
+      << "  Node = 2,\n";
+
+    unsigned nkId = 3;
+    for (auto cls : *getClasses()) t << "  " << cls->name << " = " << nkId++ << ",\n";
```

Comment:
Please see: https://github.com/p4lang/p4c/pull/4377#issuecomment-1928778499

I'd like to extend the NK_* identifiers to include the namespace hierarchy to avoid redefinition errors when there are two classes with the same name but different namespaces.

Reply:
How does this look like in .def?

Reply:
Nevermind, was able to reproduce this.